### ISSUE TITLE
[charts] Store voronoi graph in the data space not the rendering one

### ISSUE THREAD

**Issue by alexfauquette** [mui/mui-x#15737]
Currently, the voronoi is computed in the rendering space. Which means each time a user zoom, the scale is updated and we need to compute back the voronoi graph.

Instead we should compute in the data space and call the `scale.invert()` to map mouse coordinates to data values.

**Search keywords**:

**Comment by bernardobelchior** [comment]
I was taking a look at this issue and a couple of questions came up.

The goal seems to be avoiding re-creating the `Delaunay` class on zoom. To do that, we'd need to call `delaunay.find()` with the original coordinates, rather than with pixels. However, from a user point of view, we want to obtain the point that is closest to the current pointer position, which must be measured in pixels. As such, if we want to create a `Delaunay` class with the original coordinates instead of pixels, then we will need to create one `Delaunay` per combination of X axis and Y axis. This is required because we will need to transform the pointer position in pixels into the original coordinates, and that transformation depends on the scales of the X and Y axes. 

From there, for each X, Y axes combination, we'll find the closest point using `delaunay.find()`. Next, we need to check which of these points is closest **in pixels**, so we'll need to apply the respective X and Y scales to each of the closest points and find the one that is closest to the current pointer position.

Finally, we check if the closest point, in pixels, is within the Voronoi max radius as we do now.

Does this sound good? Do you see any potential for improvement? I don't love that we're creating a new `Delaunay` per X,Y axes combination, but I can't think of another solution that is correct.

-----

Another point is the type of scales.

It seems we support a set of D3 scales. Band scales are part of this set, but they don't have an invert function. However, [there seems to be a way](https://stackoverflow.com/questions/38633082/d3-getting-invert-value-of-band-scales) of implementing an invert-like function for band scales. 

Nevertheless, this raises a couple of questions:
1. Is there a use case for `ChartsVoronoiHandler` other than scatter charts? I couldn't see any usage in the codebase;
2. If there isn't, does it even make sense to have a scatter chart with a band scale? (Not sure what the implications of the answer to this question would be, to be honest)

**Comment by alexfauquette** [comment]
I was surprised I took such a weird decision as doing all the compûtation in the px dimensions and not the value, but not having to handle those edge cases of  band, points (but also Date) scale migth be the reasons.

From the various points you mentioned, I end up thinking about the following issue: If we have multiple y-axis and zoom on one of them, the result will need to be updated.


Since it looks complex, we might reconsider the goal. We could investigate following points

- Estimate the cost of doing voronoid computation on large dataset (what append on 1k-10k points ? how does it scale with the number of point (Nlog(N) in theory but better to test))
- when does it run. I assume each zoom update ruens the voronoid


Depending on those results, maybe the solution is not to revert the space of commputation, but block the voronoid update while the zoom is active. Having an update when zoom end instead of one at each step coudl already be an improvement.

**Comment by bernardobelchior** [comment]
> when does it run. I assume each zoom update ruens the voronoid

Yeah, I checked and we're creating an instance of the `Delauney` class on every zoom and on every pan.

**Comment by bernardobelchior** [comment]
I was thinking about this again and thought of documenting this for future reference:

Couldn't we create the Devaunay with zoom = 1 and then scale the results? Because zoom applies to all series, right? So in theory the point that is closest is always closest regardless of zoom level. 
We'd need to check for the max radius with the scaled coordinates, but that way we could create the Devaunay just once. Unfortunately, this does mean we'd be seaching the Devaunay for points outside the visible area, so I'm not sure how that would affect performance.

Concluding, if creating the Devaunay is more costly than searching it, then I think this approach could work. 
An improvement to improve search performance would be applying a limit to the `find` method of `Devaunay` to ensure we wouldn't search for points outside currently visible area, but not sure if this is easy without forking the `Devaunay` class.

**Comment by bernardobelchior** [comment]
I was thinking about this once more. The comment above works well for linear scales, but for log scales it won't because depending on the position on the chart, a 10px distance on the screen may result in different distances (not proportional) in the data space. 

In summary, if we want to move the computation to the data space, we need to invert the results before computing the distance. 

Would that be more performant than re-creating the Delaunay? Not sure

**Comment by alexfauquette** [comment]
I agree it does not look feasible to improve performance that way. I'm closing the issue
